plugins/search-indexer-backend/src/registry.ts

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2021 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  IndexableDocumentCollator,
+  IndexableDocumentDecorator,
+  RegisterCollatorParameters,
+  RegisterDecoratorParameters,
+} from './types';
+
+interface CollatorRegistryEntry {
+  collate: IndexableDocumentCollator;
+  refreshInterval: number;
+}
+
+export class Registry {
+  private collators: Record<string, CollatorRegistryEntry>;
+  private decorators: Record<string, IndexableDocumentDecorator[]>;
+
+  private static instance: Registry;
+
+  private constructor() {
+    this.collators = {};
+    this.decorators = {};
+  }
+
+  static getInstance(): Registry {
+    if (!Registry.instance) {
+      Registry.instance = new Registry();
+    }
+    return Registry.instance;
+  }
+
+  addCollator({
+    type,
+    collator,
+    defaultRefreshIntervalSeconds,
+  }: RegisterCollatorParameters): void {
+    this.collators[type] = {
+      refreshInterval: defaultRefreshIntervalSeconds,
+      collate: collator,
+    };
+  }
+
+  addDecorator({
+    types = ['*'],
+    decorator,
+  }: RegisterDecoratorParameters): void {
+    types.forEach(type => {
+      if (this.decorators.hasOwnProperty(type)) {
+        this.decorators[type].push(decorator);
+      } else {
+        this.decorators[type] = [decorator];
+      }
+    });
+  }
+
+  // @todo But like with coordination, timing, error handling, and what have you.
+  async execute(): Promise<void> {
+    return new Promise(resolve => {
```

Comment:
Use of the promise constructor seems a bit strange here, could prolly just rely on await and possible Promise.all with a possible limiter?